File: 948-bag-of-tokens/948-bag-of-tokens.py
class Solution:
    def bagOfTokensScore(self, tokens: List[int], power: int) -> int:
        tokens.sort()
        tokens = deque(tokens)
        score = 0
        max_score = 0
        while tokens:
            if tokens[0] <= power:
                power -= tokens.popleft()
                score += 1
                max_score = max(max_score, score)
            elif score >= 1:
                power += tokens.pop()
                score -= 1
            else:
                break
        return max_score

File: 948-bag-of-tokens/948-bag-of-tokens.cpp
class Solution {
public:
    int bagOfTokensScore(vector<int>& tokens, int power) {
        sort(tokens.begin(), tokens.end());
        deque<int> d;
        for (int i=0; i<tokens.size(); i++) {
            d.push_back(tokens[i]);
        }
        int score = 0;
        int max_score = 0;
        while (!d.empty()) {
            if (d.front() <= power) {
                power -= d.front();
                d.pop_front();
                score++;
                max_score = std::max(max_score, score);
            }
            else {
                if (score >= 1) {
                    power += d.back();
                    d.pop_back();
                    score--;
                }
                else {
                    break;
                }
            }
        }
        return max_score;
    }
};
